ath79: add support for Fortinet FAP-221-C

StereoRocker opened the pull request:
Fortinet FAP-221-C is an indoor access point with 1Gb ethernet port,
dual-band wireless, internal antenna plates, and PoE.

Hardware and board design from Senao, although this variant uses a 32MB
flash chip whereas the base design appears to use a 16MB chip.

# Specifications
    SoC:        Qualcomm Atheros QCA9558
    WiFi 1:     Qualcomm Atheros QCA9558
    WiFi 2:     Qualcomm Atheros QCA9882
    Ethernet:   Qualcomm Atheros QCA9558 (1x GbE, 802.3af PoE)
    RAM:        256MB
    Flash:      32MB
	Buttons:    Single reset button (GPIO)
    LEDs:       WiFi 2 (r/g), WiFi 1 (r/g), LAN (r/g)*, Diag (r/g),
                Power (g)
                * LAN LED behaves odd. Red seems to be hardwired to
                  ethernet status somehow. Green can be controlled
                  via GPIO as expected.

# UART

```
Unmarked, but populated 4-pin header

[o o o o]
 | | | ^- VCC, pin 1
 | |  `-- GND
 |  `---- TX
  `------ RX
```

The UART runs at 9600-8-N-1.

# Installation

## OEM firmware

Upload squashfs-factory.bin to a TFTP server. Connect to the device via
SSH, with the default username "admin" and no password. Run the following
command:

`restore <filename> <TFTP IP>`

The same can also be achieved by logging into the console via UART.

## u-boot

Connect to the UART console, and power on the device. Interrupt the boot
sequence when indicated. Press [G] to upload a new firmware. Enter the
details of your TFTP server, assign an available IP to the device, and
enter the name of the firmware file on the TFTP server. The file will be
transferred to the device, then press [D] to save the image to flash.

# Backup OEM firmware

Fortinet do not provide binaries for these devices to the public. So if
you ever want to go back to OEM firmware, follow these steps to dump and
generate a flashable image.

Power up the device, and identify its IP address on the network. Run the
following commands on your local machine, assuming a Linux environment.

```
# These commands dump the individual firmware partitions
ssh admin@<deviceip> "cat /dev/mtdblock1" > mtdblock1
ssh admin@<deviceip> "cat /dev/mtdblock2" > mtdblock2
ssh admin@<deviceip> "cat /dev/mtdblock3" > mtdblock3

# These commands assemble the firmware image, image.out
cat mtdblock1 mtdblock2 mtdblock3 > FP221C
gzip FP221C
mv FP221C image.out
```

The resulting file "image.out" can be flashed by u-boot. Note that due
to containing the additional "reserved" partition, the `restore` command
will not allow flashing this image back. Omit mtdblock3 to generate an
image that can be flashed with `restore`.

Signed-off-by: Dominic Houghton <technpizza@gmail.com>

hauke commented:
Please rebase this on top of master which already contains e085812a7d7022c4bec948280efbcba738101636

mpratt14 commented:
@StereoRocker what's the FCC ID?

also, if you need help with using Git to rebase as hauke requested, just ask

the reason is, the other commits are accepted already

StereoRocker committed:
ath79: add support for Fortinet FAP-221-C

Fortinet FAP-221-C is an indoor access point with 1Gb ethernet port,
dual-band wireless, internal antenna plates, and PoE.

Hardware and board design from Senao, although this variant uses a 32MB
flash chip whereas the base design appears to use a 16MB chip.

# Specifications
    SoC:        Qualcomm Atheros QCA9558
    WiFi 1:     Qualcomm Atheros QCA9558
    WiFi 2:     Qualcomm Atheros QCA9882
    Ethernet:   Qualcomm Atheros QCA9558 (1x GbE, 802.3af PoE)
    RAM:        256MB
    Flash:      32MB
	Buttons:    Single reset button (GPIO)
    LEDs:       WiFi 2 (r/g), WiFi 1 (r/g), LAN (r/g)*, Diag (r/g),
                Power (g)
                * LAN LED behaves odd. Red seems to be hardwired to
                  ethernet status somehow. Green can be controlled
                  via GPIO as expected.

# UART

```
Unmarked, but populated 4-pin header

[o o o o]
 | | | ^- VCC, pin 1
 | |  `-- GND
 |  `---- TX
  `------ RX
```

The UART runs at 9600-8-N-1.

# Installation

## OEM firmware

Upload squashfs-factory.bin to a TFTP server. Connect to the device via
SSH, with the default username "admin" and no password. Run the following
command:

`restore <filename> <TFTP IP>`

The same can also be achieved by logging into the console via UART.

## u-boot

Connect to the UART console, and power on the device. Interrupt the boot
sequence when indicated. Press [G] to upload a new firmware. Enter the
details of your TFTP server, assign an available IP to the device, and
enter the name of the firmware file on the TFTP server. The file will be
transferred to the device, then press [D] to save the image to flash.

# Backup OEM firmware

Fortinet do not provide binaries for these devices to the public. So if
you ever want to go back to OEM firmware, follow these steps to dump and
generate a flashable image.

Power up the device, and identify its IP address on the network. Run the
following commands on your local machine, assuming a Linux environment.

```
# These commands dump the individual firmware partitions
ssh admin@<deviceip> "cat /dev/mtdblock1" > mtdblock1
ssh admin@<deviceip> "cat /dev/mtdblock2" > mtdblock2
ssh admin@<deviceip> "cat /dev/mtdblock3" > mtdblock3

# These commands assemble the firmware image, image.out
cat mtdblock1 mtdblock2 mtdblock3 > FP221C
gzip FP221C
mv FP221C image.out
```

The resulting file "image.out" can be flashed by u-boot. Note that due
to containing the additional "reserved" partition, the `restore` command
will not allow flashing this image back. Omit mtdblock3 to generate an
image that can be flashed with `restore`.

Signed-off-by: Dominic Houghton <technpizza@gmail.com>

StereoRocker commented:
Apologies for the delay @hauke and thank you for the reminder @mpratt14 

I've rebased this on master now. 

The FCC ID is "TVE-121402".

mpratt14 commented:
you can note in the commit message that for Ethernet it uses AR8035-A

hardware is the same as EAP1750

mpratt14 commented:
please try to use existing DTSI for senao boards like Watchguard AP300 and Engenius EAP1750 for example

also try to use wifi calibration through the DTS instead of caldata_extract, like I just did for 221-B

StereoRocker commented:
> also try to use wifi calibration through the DTS instead of caldata_extract, like I just did for 221-B

What's the benefit of doing this?

mpratt14 commented:
they're trying to get rid of caldata_extract since it is possible to do in DTS now

it's a general goal to handle as much of device configuration within DTS since it is upstream Linux standard and that's what it's designed to do

StereoRocker commented:
Makes sense, thanks for explaining. :)

On Sun, 15 Jan 2023, 16:02 Michael Pratt, ***@***.***> wrote:

> they're trying to get rid of caldata_extract since it is possible to do in
> DTS now
>
> it's a general goal to handle as much of device configuration within DTS
> since it is upstream Linux standard and that's what it's designed to do
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/11621#issuecomment-1383187506>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHM5CKRUQN36VG7HGFSXS3WSQNRHANCNFSM6AAAAAATKODI3Y>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mpratt14 commented:
this is very picky but, if there are 3 antennas connected to the PCIe wifi chip, then it is probably QCA9880 instead of QCA9882

mpratt14 commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1340,6 +1340,23 @@ define Device/fortinet_fap-221-b
+  DEVICE_PACKAGES := ath10k-firmware-qca988x-ct kmod-ath10k-ct
+  IMAGE_SIZE := 20480k
+  LOADER_FLASH_OFFS := 0x0040000
+  IMAGE/factory.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | \
+	  append-rootfs | pad-rootfs | \
+	  check-size | pad-to $$$$(IMAGE_SIZE) | \
+	  append-loader-okli-uimage $(1) | pad-to 10944k | \
+	  gzip-filename $$$$(FACTORY_IMG_NAME)
```
the last `pad-to` must also be adjusted for the same reason that `IMAGE_SIZE` was adjusted for 32 MB

this last padding is so that the size of the image is equivalent to the full size of the partitions together. If it does not match up to the full size, the OEM upgrade will not write the last partition.

for 221-B this was 10944k because:
```
(number of blocks in each partition in firmware)
fwconcat0 --> 0x90 --> 144
loader --> 0x01 --> 1
fwconcat1 --> 0x1a --> 26

(64k size of each block)
144 + 1 + 26 = 171
171 * 64k = 10944k
```

this is also the same as the OEM `kernel` + `rootfs`

mpratt14 replied:
in this case:

```
fwconcat0 --> 0x140
loader --> 0x001
fwconcat1 --> 0x03f
```

mpratt14 commented:
@StereoRocker do you need help or are you busy?

StereoRocker commented:
Just busy, it'll be a couple weeks before I can pick this up again. Thanks
for asking. :)

On Sat, 4 Feb 2023, 18:20 Michael Pratt, ***@***.***> wrote:

> @StereoRocker <https://github.com/StereoRocker> do you need help or are
> you busy?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/11621#issuecomment-1416817966>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHM5CLQ2ECPOB6CI2FOCY3WV2MWNANCNFSM6AAAAAATKODI3Y>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Leo-PL commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:
```diff
@@ -63,6 +63,9 @@ case "$FIRMWARE" in
 		ath9k_patch_mac $(mtd_get_mac_ascii cfg1 RADIOADDR1)
 		;;
+	fortinet,fap-221-c)
+		caldata_extract "art" 0x1000 0x440
```
Please extract caldata for ath10k the same way as for ath9k, by device tree bindings.

Leo-PL commented on target/linux/ath79/generic/base-files/lib/preinit/10_fix_eth_mac.sh:
```diff
@@ -19,7 +19,8 @@ preinit_set_mac_address() {
-	fortinet,fap-221-b)
+	fortinet,fap-221-b|\
+	fortinet,fap-221-c)
 		ip link set dev eth0 address $(mtd_get_mac_text u-boot 0x3ff80 12)
```
Please use nvmem-layout for MAC extraction. See FAP-220-B for example.

hauke commented:
Support for this device was merged in #18109
See 6da7352ebe1150928a249aff47eb28c5c81f26aa
